### PR TITLE
When recovering from a `:` in a pattern, use adequate AST pattern

### PR DESCRIPTION
**Pull request by estebank**
If the suggestion to use `::` instead of `:` in the pattern isn't correct, a second resolution error will be emitted.

### CODE REVIEW

**Comment by estebank** [comment]
ping

**Review by nagisa — COMMENTED** [review]
I think extracting the code to "join" two patterns into a function might make the code cleaner.

File: compiler/rustc_parse/src/parser/pat.rs [hunk 1]
```diff
@@ -126,17 +128,84 @@ impl<'a> Parser<'a> {
...
+                                                        0,
+                                                        PathSegment::from_ident(ident.clone()),
+                                                    );
+                                                    path.span = new_span;
+                                                    show_sugg = true;
+                                                    first_pat = pat;
+                                                }
+                                                PatKind::Path(old_qself, old_path) => {
+                                                    for segment in old_path.segments.iter().rev() {
+                                                        path.segments.insert(0, segment.clone());
```
[the remark this answers is not in the document]
Comment (estebank) [reply compiler/rustc_parse/src/parser/pat.rs:151]: What do you think of this?

```rust
                                                    path.segments = old_path
                                                        .segments
                                                        .iter()
                                                        .cloned()
                                                        .chain(take(&mut path.segments))
                                                        .collect();
```

File: src/test/ui/parser/issue-87086-colon-path-sep.rs [hunk 2]
```diff
@@ -16,40 +20,40 @@ fn g1() {
...
+    if let Foo:Bar = f() {
     //~^ ERROR: expected one of
     //~| HELP: maybe write a path separator here
     }
 }
 
 fn g1_neg() {
     match f() {
-        ref Foo: Bar::Baz => {}
+        ref qux: Foo::Baz => {}
```
[the remark this answers is not in the document]
Comment (estebank) [reply src/test/ui/parser/issue-87086-colon-path-sep.rs:48]: This wasn't *intended* as a negative test, just making sure it was covered. It doesn't reach the recovery code. Looking.

[the remark this answers is not in the document]
Comment (estebank) [reply src/test/ui/parser/issue-87086-colon-path-sep.rs]: Fixed

File: compiler/rustc_parse/src/parser/diagnostics.rs [hunk 3]
```diff
@@ -2075,4 +2080,177 @@ impl<'a> Parser<'a> {
...
+                                }
+                            }
+                            _ => {}
+                        }
+                        if show_sugg {
+                            err.span_suggestion(
+                                span,
+                                "maybe write a path separator here",
+                                "::".to_string(),
+                                Applicability::MachineApplicable,
```
Comment (nagisa) [inline compiler/rustc_parse/src/parser/diagnostics.rs:2185]: Is this truly `MachineApplicable`? We don't really check with the resolver here that the constructed pattern will match, do we?

**Review by nagisa — APPROVED** [review]
Seems good, except for machine applicability stuff.

**Pull request merged**